python/analyzers/HNLAnalyzer.py

```diff
@@ -408,22 +500,94 @@ def process(self, event):
+        event.the_3lep_cand.abs_tot_iso04_deltaBeta  = totIso(event, 'dBeta', 0.4) 
+        event.the_3lep_cand.abs_tot_iso05_deltaBeta  = totIso(event, 'dBeta', 0.5) 
 
+        event.the_3lep_cand.rel_tot_iso03_deltaBeta  =  event.the_3lep_cand.abs_tot_iso03_deltaBeta /  event.the_3lep_cand.hnVisP4().pt()
+        event.the_3lep_cand.rel_tot_iso04_deltaBeta  =  event.the_3lep_cand.abs_tot_iso04_deltaBeta /  event.the_3lep_cand.hnVisP4().pt()
+        event.the_3lep_cand.rel_tot_iso05_deltaBeta  =  event.the_3lep_cand.abs_tot_iso05_deltaBeta /  event.the_3lep_cand.hnVisP4().pt()
+
+
+        #####################################################################################
+        # After passing all selections and we have an HNL candidate, pass a "true" boolean!
+        #####################################################################################
         return True
-        
-        
+
+
+
+def totIso(event, offset_mode, dRCone):
+    ch_pu_iso = chargedHadronIso(event, dRCone, True)
+    ch_pv_iso = chargedHadronIso(event, dRCone, False)
+    neu_iso   = neutralHadronIso(event, dRCone)
+    ph_iso    = photonIso(event, dRCone)
+    if offset_mode == 'rhoArea': 
+        eta = event.the_3lep_cand.hnVisP4().eta()
+        offset = offset_rhoArea(event.rho, dRCone, eta)
+    if offset_mode == 'dBeta': 
+        offset = offset_dBeta(0.5, ch_pu_iso)
+    tot_iso = ch_pv_iso + max(0., ph_iso + neu_iso - offset)
+    # if dRCone == 0.3:
+        # print '2M dr %.1f: ch_pv_iso: %.2f, neu_iso: %.2f, ph_iso: %.2f, ch_pu_iso: %.2f, l1+l2pt: %.2f, id: %i'%(dRCone, ch_pv_iso, neu_iso, ph_iso, ch_pu_iso, event.the_3lep_cand.l1().pt()+event.the_3lep_cand.l2().pt(), event.eventId)
+    return tot_iso
+
+def offset_rhoArea(rho, dRCone, eta):
+    area = 0.0
+    if abs(eta) < 0.8000: area = 0.0566
+    if abs(eta) > 0.8000 and abs(eta) < 1.3000: area = 0.0562
+    if abs(eta) > 1.3000 and abs(eta) < 2.0000: area = 0.0363
+    if abs(eta) > 2.0000 and abs(eta) < 2.2000: area = 0.0119
+    if abs(eta) > 2.2000 and abs(eta) < 2.4000: area = 0.0064
+    if dRCone != 0.3: area *= ( (dRCone ** 2) / (0.3 **2) )
+    # print 'area = {a}, offset = {o}'.format(a = area, o = area * rho) 
+    return area * rho
     
+def offset_dBeta(dBeta, ch_pu_iso):
+    # print 'dbeta = {db}, offset = {o}'.format(db = dBeta, o = dBeta * ch_pu_iso)
+    return ch_pu_iso * dBeta
+        
+def chargedHadronIso(event, dRCone, PU = False): 
+    if PU == True:
+        charged_pfs = [ipf for ipf in event.pfs if ( ipf.charge() != 0 and ipf.pt() > 0.5 )]
+        charged_pfs = [ipf for ipf in charged_pfs if ipf.fromPV() <= 1]
+    if PU == False:
+        charged_pfs = [ipf for ipf in event.pfs if ( ipf.charge() != 0 and abs(ipf.pdgId()) != 11 and abs(ipf.pdgId()) != 13 and ipf.pt() > 0.5 )]
+        charged_pfs = [ipf for ipf in charged_pfs if (ipf.fromPV() > 1 and abs(ipf.pdgId()) == 211) ]
```

Comment:
same quesiton on `fromPV`.
I think that you require pdg ID 211 to exclude electrons and muons from the isolation sum. But that's done above at L572  already, isn't it?

Reply:
this i unfortunately can't answer, maybe @vinzenzstampf can comment on this?

Reply:
@vinzenzstampf can you comment on this?

Reply:
@dehuazhu can you knock on Vinz's door?

Reply:
still nothing on this?